### PR TITLE
Basic regression example using MLJFlux

### PR DESCRIPTION
**Pull request by ayush-1506**
Basic regression (using the boston dataset), using MLJFlux's `NeuralNetworkRegressor`.

### CODE REVIEW

**Comment by tlienart** [comment]
Awesome thanks, I'll have a look at this this week

**Comment by ablaom** [comment]
@ayush-1506 Thanks for this! However, I suggest we hold out releasing this until we have registered a release, and have more complete docs. 

There is still [this open issue](https://github.com/alan-turing-institute/MLJFlux.jl/issues/31) and [this unmarked PR](https://github.com/alan-turing-institute/MLJFlux.jl/pull/33) related to images.  I think we should holdout on a release until the image classifier is sorted. This is the main use-case.

**Comment by ayush-1506** [comment]
Would love a review. I've updated the example to accommodate the latest changes in MLJFlux. Let me know it you'd like me to add another (fancier) example (probably something to do with images)

@ablaom @tlienart .

**Comment by azev77** [comment]
The classics: MNIST & CIFAR?

File: _literate/EX-boston-flux.jl [hunk 1]
```diff
@@ -0,0 +1,128 @@
+# **Main author**: Ayush Shridhar (ayush-1506).
+#
+# ## Getting started
+
+using MLJFlux, MLJ, DataFrames, Statistics, Flux
```
Comment (tlienart) [inline _literate/EX-boston-flux.jl:5]: please `import` specific names from `DataFrames` to avoid name clashes with MLJ

also you seem to use qualified imports from Flux and MLJFlux so might as well just `import` them instead of `using`

File: _literate/EX-boston-flux.jl [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    n1::Int #Number of cells in the first hidden layer
+    n2::Int #Number of cells in the second hidden layer
+end
+
+# Step 2: Building the neural network from this object.
+# Extend the `MLJFlux.fit` function. This takes in 3 arguments: The object of
+# `MyNNRegressor`, input dimension (ip) and output dimension (op).
+
+function MLJFlux.build(model::MyNNRegressor, ip, op)
+    return Flux.Chain(Flux.Dense(ip, model.n1), Flux.Dense(model.n1, model.n2), Flux.Dense(model.n2, op))
```
Comment (tlienart) [inline _literate/EX-boston-flux.jl:47]: for readability, I'd rather have `layer1 = ...`, `layer2= ...`, `layer_out = ...` and then chain. Also just use more explicit names `input_dims` `output_dims`

File: _literate/EX-boston-flux.jl [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+nnregressor = NeuralNetworkRegressor(builder=myregressor, epochs=10)
+
+# Other parameters that `NeuralNetworkRegressor` takes in are:
+# optimiser = ADAM(0.001, (0.9, 0.999), IdDict{Any,Any}())
+# loss = Flux.mse,
+# epochs = 10,
+# batch_size = 1,
+# lambda = 0.0,
+# alpha = 0.0,
+# optimiser_changes_trigger_retraining = false
```
Comment (tlienart) [inline _literate/EX-boston-flux.jl:72]: (1) add `*` in front of each line so that it appears as a list
(2) put backticks to make it look like code
(3) add a short explanation of what it's about 
(4) link to docs and/or readme of mljflux for more details?

File: _literate/EX-boston-flux.jl [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+mach = machine(nnregressor, features, targets)
+
+# Let's fit this on the train set
+
+fit!(mach, rows=train, verbosity=3)
+
+# As we can see, the loss decreases at each epoch, showing the the neural network 
+# is gradually learning form the training set.
+
+predict(mach, rows=test)
```
Comment (tlienart) [inline _literate/EX-boston-flux.jl:86]: so if you just leave this like that, the full output will be displayed on the webpage. Either add a `;` or give that value to a variable and show the first few values?

File: _literate/EX-boston-flux.jl (hunk 1 above)
Comment (ablaom) [reply _literate/EX-boston-flux.jl:5]: You need `using Random` here.

File: _literate/EX-boston-flux.jl [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+# Next obvious steps: partitioning into train and test set
+
+train, test = partition(eachindex(targets), 0.70, shuffle=true, rng=52)
+
+# Let us try to implement an Neural Network regressor using
+# Flux.jl. MLJFlux.jl provides an MLJ interface to the Flux.jl
+# deep learning framework. The package provides four essential
+# models: `NeuralNetworkRegressor, MultivariateNeuralNetworkRegressor,
+# NeuralNetworkClassifier` and `ImageClassifier`.
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:28]: Replace `MulitvariateNeural...` with `MultitargetNeural...`

File: _literate/EX-boston-flux.jl [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+# Let us try to implement an Neural Network regressor using
+# Flux.jl. MLJFlux.jl provides an MLJ interface to the Flux.jl
+# deep learning framework. The package provides four essential
+# models: `NeuralNetworkRegressor, MultivariateNeuralNetworkRegressor,
+# NeuralNetworkClassifier` and `ImageClassifier`. 
+
+# At the heart of these models is a neural network. This is specified using
+# the `builder` parameter. Creating a builder object consists of two steps:
+# 1. Creating a new struct inherited from `MLJFlux.Builder`. `MLJFlux.Builder`
+# is an abstract structure used for the purpose of dispatching. Suppose we define
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:33]: (FYI: We no longer require the struct to subtype `MLJBuilder`, but if it does then the `show` will be consistent with MLJ.)

File: _literate/EX-boston-flux.jl [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+# build the model later. (Step 2). Let's use Dense Neural Network with 2 hidden layers.
+
+mutable struct MyNNRegressor <: MLJFlux.Builder
+    n1::Int #Number of cells in the first hidden layer
+    n2::Int #Number of cells in the second hidden layer
+end
+
+# Step 2: Building the neural network from this object.
+# Extend the `MLJFlux.fit` function. This takes in 3 arguments: The object of
+# `MyNNRegressor`, input dimension (ip) and output dimension (op).
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:44]: `fit` -> `build`

File: _literate/EX-boston-flux.jl [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+function MLJFlux.fit(model::MyNNRegressor, ip, op)
+    return Flux.Chain(Flux.Dense(ip, model.n1), Flux.Dense(model.n1, model.n2), Flux.Dense(model.n2, op))
+end
+
+# With all definitions ready, let us create an oject of this:
+
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:51]: typo "object" -> "object"

File: _literate/EX-boston-flux.jl [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+features, targets = @load_boston
+features = DataFrame(features)
+@show size(features)
+@show targets[1:3]
+first(features, 3) |> pretty
+
+# Next obvious steps: partitioning into train and test set
+
+train, test = partition(eachindex(targets), 0.70, shuffle=true, rng=52)
+
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:23]: if you specify `rng` then `shuffle=true` is redundant.

File: _literate/EX-boston-flux.jl [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+# the `builder` parameter. Creating a builder object consists of two steps:
+# 1. Creating a new struct inherited from `MLJFlux.Builder`. `MLJFlux.Builder`
+# is an abstract structure used for the purpose of dispatching. Suppose we define
+# a new struct called `MyNNRegressor`. This can contain any attribute required to
+# build the model later. (Step 2). Let's use Dense Neural Network with 2 hidden layers.
+
+mutable struct MyNNRegressor <: MLJFlux.Builder
+    n1::Int #Number of cells in the first hidden layer
+    n2::Int #Number of cells in the second hidden layer
+end
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:40]: I think it's confusing to name the builder object "MyNNRegressor". In principle the same builder could be used for classification, and so forth. How about "MyBuilder" or "MyNetworkBuilder"?

File: _literate/EX-boston-flux.jl [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+# With all definitions ready, let us create an oject of this:
+
+myregressor = MyNNRegressor(20, 10)
+
+# Since the boston dataset is a regression problem, we'll be using
+# `NeuralNetworkRegressor` here. One thing to remember is that 
+# a `NeuralNetworkRegressor` object works seamlessly like any other
+# MLJ model: you can wrap it in an  MLJ `machine` and to anything
+# you'd do otherwise.
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:58]: typo "and to anything" -> "and do anything"

File: _literate/EX-boston-flux.jl [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+nnregressor = NeuralNetworkRegressor(builder=myregressor, epochs=10)
+
+# Other parameters that `NeuralNetworkRegressor` takes in are:
+# optimiser = ADAM(0.001, (0.9, 0.999), IdDict{Any,Any}())
+# loss = Flux.mse,
+# epochs = 10,
+# batch_size = 1,
+# lambda = 0.0,
+# alpha = 0.0,
+# optimiser_changes_trigger_retraining = false
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:72]: This list seems redundant as it just repeats the output of if the `show`. More useful would be a description of what each parameter means, which you can copy from the the readme: 

https://github.com/alan-turing-institute/MLJFlux.jl#model-hyperparameters

File: _literate/EX-boston-flux.jl [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+# `nnregressor` now acts like any other MLJ model. Let's try wrapping it in a 
+# MLJ machine and calling `fit!, predict!`.
+
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:76]: typo "predict!" -> "predict"

File: _literate/EX-boston-flux.jl [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+mach = machine(nnregressor, features, targets)
+
+# Let's fit this on the train set
+
+fit!(mach, 3, rows=train, vebosity=3)
+
+# As we can see, the loss decreases at each epoch, showing the the neural network 
+# is gradually learning form the training set.
+
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:85]: Please replace "loss" with "training loss"

File: _literate/EX-boston-flux.jl [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+# parameter. (Look at documentation for `fit!` for more).
+
+# However, changing parameters such as batch_size will necessarily cause re-training from scratch.
+nnregressor.batch_size = 2
+fit!(mach, rows=train, verbosity=3)
+
+# Another bit to remember here is that changing the optimiser doesn't cause retaining by default.
+# However, the `optimiser_changes_trigger_retraining` in NeuralNetworkRegressor can be toggled to
+# accomodate this.
+
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:107]: Add something like, "This allows one to modify the learning rate, for example, after an initial burn-in period".

File: _literate/EX-boston-flux.jl [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+## Tuning
+
+# As mentioned above, `nnregressor` can act like any other MLJ model. Let's try to tune the
+# batch_size parameter.
+
+bs = range(nnregressor, :batch_size, lower=1, upper=5)
+
+tm = TunedModel(model=nnregressor, ranges=[bs, ], measure=Flux.mse)
+
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:116]: Unfortunately, I don't believe the Flux and MLJ API's for loss functions are compatible. See the note at the end of the readme. So here, you should use `measure=l2`, say.

**Comment by ablaom** [comment]
One last comment: Rather than `predict(mach, rows=test)` specify the data explicitly, as in `predict(mach, X[test, :])`. Otherwise the reader may wonder, How do I predict on new data?

**Comment by ayush-1506** [comment]
Thanks for the reviews! I'm traveling right now so I'll get to this as soon as I've reached my destination.

**Comment by ablaom** [comment]
Bon voyage. 

After https://github.com/JuliaRegistries/General/pull/16542 is merged I'll update the MLJ model registry and you can use `@load NeuralNetworkRegressor` to load the model (no need for explicit `using MLJFlux`).

**Comment by ayush-1506** [comment]
Awesome, I'll update the PR with this too once JuliaRegistries/General#16542 gets merged.

**Comment by ablaom** [comment]
https://github.com/JuliaRegistries/General/pull/16728

**Comment by ayush-1506** [comment]
`add MLJFlux` now gives me an error:
```
ERROR: Unsatisfiable requirements detected for package MLJFlux [094fc8d1]:
 MLJFlux [094fc8d1] log:
 ├─possible versions are: 0.1.0-0.1.1 or uninstalled
 ├─restricted to versions * by an explicit requirement, leaving only versions 0.1.0-0.1.1
 └─restricted by compatibility requirements with MLJModelInterface [e80e1ace] to versions: uninstalled — no versions left
   └─MLJModelInterface [e80e1ace] log:
     ├─possible versions are: [0.1.0-0.1.8, 0.2.0-0.2.7, 0.3.0] or uninstalled
     └─restricted to versions 0.2.7 by an explicit requirement, leaving only versions 0.2.7
```
Investigating which requirement is causing this.

**Comment by ayush-1506** [comment]
@abloam Turns out I was on an older commit and a simple rebase on top of master fixes this. I got another version issue with lossfunctions on doing `]add MLJFlux` and https://github.com/alan-turing-institute/MLJFlux.jl/pull/50 should fix that.

**Comment by ablaom** [comment]
https://github.com/JuliaRegistries/General/pull/16944

**Comment by ablaom** [comment]
@ayush-1506 Are you able to resolve this conflict or do you need help?

**Comment by ayush-1506** [comment]
Resolved: Didn't realize newer commits were causing issues.

**Comment by tlienart** [comment]
is that good to merge? @ablaom @ayush-1506

File: _literate/EX-boston-flux.jl [hunk 17]
```diff
@@ -92,37 +97,50 @@ predict(mach, rows=test)
...
-# accomodate this.
+# accomodate this. This allows one to modify the learning rate, for example, after an initial burn-in period.
+
+## Inspecting out-of-sample loss as a function of epochs
+
+r = MLJ.range(nnregressor, :epochs, lower=1, upper=30, scale=:log10)
+curve = MLJ.learning_curve(nnregressor, features, targets,
+                       range=r,
+                       resampling=MLJ.Holdout(fraction_train=0.7),
+                       measure=MLJ.rms)
```
Comment (ablaom) [inline _literate/EX-boston-flux.jl:119]: For consistency with the tuning below, make this `measure=MLJ.l2`.

**Comment by ablaom** [comment]
Okay, just reviewed the revised content. Just one small detail. There is an inconsistent choice of MLJ measure: in the learning curve it is `rms` but in the tuning exercise it is `l2`. I've noted it just above. 

Otherwise good to go.

(I've not reviewed for formatting)

**Comment by ayush-1506** [comment]
Thanks for the reviews @ablaom @tlienart .I believe this can be merged now. (The plan is to add another example of using MLJFlux that also shows multiple levels of stacking, but that is probably going to be a new PR)
